pkg/nsx/services/builder.go

```diff
@@ -292,190 +300,172 @@ func (service *SecurityPolicyService) buildPolicyGroupPath(obj *v1alpha1.Securit
 }
 
 func (service *SecurityPolicyService) buildRuleAndGroups(obj *v1alpha1.SecurityPolicy,
-	rule *v1alpha1.SecurityPolicyRule, ruleIdx int) (*model.Rule, *[]model.Group, error) {
-	sequenceNumber := int64(ruleIdx)
-	nsxRuleID := service.buildRuleID(obj, ruleIdx)
-	var nsxRuleName string
-	var ruleGroups []model.Group
+	rule *v1alpha1.SecurityPolicyRule, ruleIdx int,
+) ([]*model.Rule, []*model.Group, error) {
+	var ruleGroups []*model.Group
 	var nsxRuleAppliedGroup *model.Group
 	var nsxRuleSrcGroup *model.Group
 	var nsxRuleDstGroup *model.Group
 	var nsxRuleAppliedGroupPath string
 	var nsxRuleDstGroupPath string
 	var nsxRuleSrcGroupPath string
-	var err error = nil
+	var err error
 
-	direction, err := getRuleDirection(rule)
+	ruleDirection, err := getRuleDirection(rule)
 	if err != nil {
 		return nil, nil, err
 	}
-	ruleAction, err := getRuleAction(rule)
+
+	// Since a named port may map to multiple port numbers, then it would return multiple rules.
+	// We use the destination port number of service entry to group the rules.
+	ipSetGroups, nsxRules, err := service.expandRule(obj, rule, ruleIdx)
 	if err != nil {
 		return nil, nil, err
 	}
-
-	if len(rule.Name) > 0 {
-		nsxRuleName = rule.Name
-	} else {
-		nsxRuleName = fmt.Sprintf("%s-%d", obj.ObjectMeta.Name, ruleIdx)
-	}
-
-	nsxRule := model.Rule{
-		Id:             &nsxRuleID,
-		DisplayName:    &nsxRuleName,
-		Direction:      &direction,
-		SequenceNumber: &sequenceNumber,
-		Action:         &ruleAction,
-		Services:       []string{"ANY"},
-		Tags:           service.buildBasicTags(obj),
+	for _, g := range ipSetGroups {
+		ruleGroups = append(ruleGroups, g)
```

Comment:
Do we need put ipSetGroups in RuleGroups? To my understanding,  the generated ipSetGroups will be put in the rule's destination group path, and we will handle it in the buildRuleInGroup and buildRuleOutGroup. So this step is necessary?

Reply:
Yes, all the groups will be returned to `buildSecurityPolicy`, and to be handled by nsx-t CRUD.